update resource management docs keep align with latest SDK (#30177)
dotnet/docs: docs/azure/sdk/resource-management.md

The next-generation of .NET SDK's management (or "management plane") libraries will help you create, provision, and manage Azure resources from within .NET applications. All Azure services have corresponding management libraries.

> With the management libraries (namespace beginning with `Azure.ResourceManager`, for example, `Azure.ResourceManager.Compute`), you can write configuration and deployment programs to perform the same tasks that you can through the Azure portal, Azure CLI, or other resource management tools.

Reviewer: ```suggestion
With the management libraries (namespaces beginning with `Azure.ResourceManager`, for example, `Azure.ResourceManager.Compute`), you can write configuration and deployment programs to perform the same tasks that you can through the Azure portal, Azure CLI, or other resource management tools.
```

- Distributed tracing.

> [!NOTE]

> > You may notice that some packages are still in beta version, a phased releases of additional Azure services' management plane libraries are in process. If you are looking for a stable version package for a particular azure resource and currently only beta version is available, please raise an issue in [Azure SDK For .Net Github repo](https://github.com/Azure/azure-sdk-for-net/issues/new?assignees=&labels=&template=02_feature_request.yml&title=%5BFEATURE+REQ%5D)

Reviewer: ```suggestion
> You may notice that some packages are still preleased versions, phased releases of additional Azure services' management plane libraries are in process. If you are looking for a stable version package for a particular azure resource and currently only a release version is available, please raise an issue in [Azure SDK for .NET Github repo](https://github.com/Azure/azure-sdk-for-net/issues/new?assignees=&labels=&template=02_feature_request.yml&title=%5BFEATURE+REQ%5D)
```

// Code omitted for brevity

ArmClient armClient = new ArmClient(new DefaultAzureCredential());
```

For more information about the `Azure.Identity.DefaultAzureCredential` class, see [DefaultAzureCredential Class](/dotnet/api/azure.identity.defaultazurecredential).

### Management SDK Cheat Sheet

> To quickly get start with our new version Azure management SDK for .NET in less than 5 minutes, imagin you have a task to create/list/update/delete your service bus namespace(a typical kind of Azure resource), below is all the code needed. Just two steps needed.

Reviewer 2: ```suggestion
To quickly get start with our new version Azure management SDK for .NET in less than 5 minutes, imagine you have a task to create/list/update/delete your service bus namespace(a typical kind of Azure resource), below is all the code needed. Just two steps needed.
```

> To quickly get start with our new version Azure management SDK for .NET in less than 5 minutes, imagin you have a task to create/list/update/delete your service bus namespace(a typical kind of Azure resource), below is all the code needed. Just two steps needed.

Reviewer: ```suggestion
To get started with the Azure management SDK for .NET, imagine you have a task to create/list/update/delete a typical Azure service bus namespace, follow these steps:
```

First, authenticate and select the subscription & resource group that you want to work on.
```csharp
using Azure.Identity;
using Azure.ResourceManager;
using Azure.ResourceManager.ServiceBus;

ArmClient armClient = new ArmClient(new DefaultAzureCredential());
SubscriptionResource subscription = armClient.GetDefaultSubscription();

> ResourceGroupResource resourceGroupResource = client.GetDefaultSubscription().GetResourceGroup(resourceGroupName).Value;

Reviewer 2: Since we have the implicit operator between `Response<T>` and `T`, actually we could omit the `.Value` here if it does not cause confusion

> ArmClient armClient = new ArmClient(new DefaultAzureCredential());
> SubscriptionResource subscription = armClient.GetDefaultSubscription();
> ResourceGroupResource resourceGroupResource = client.GetDefaultSubscription().GetResourceGroup(resourceGroupName).Value;

Reviewer: ```suggestion
var armClient = new ArmClient(new DefaultAzureCredential());
SubscriptionResource subscription = armClient.GetDefaultSubscription();
ResourceGroupResource resourceGroupResource = 
    client.GetDefaultSubscription().GetResourceGroup(resourceGroupName);
```

```
Second, find the corresponding method to manage your Azure resource
| Operation | Method |
|-|-|

> | Get a resource with resource identifier | armClient.GetServiceBusQueueResource(ResourceIdentifier resourceIdentifier) |

Reviewer 2: We should quote all code literals
```suggestion
| Get a resource with resource identifier | `armClient.GetServiceBusQueueResource(ResourceIdentifier resourceIdentifier)` |
```

> | List| resourceGroupResource.GetServiceBusNamespaces().GetAll() |

Reviewer 2: Well, given the name and the interface indicates, the collection supports to be iterate over - the suffix "collection" and it implements `IEnumerable<T>`, therefore we could simplify this to 
```suggestion
| List| resourceGroupResource.GetServiceBusNamespaces() |
```
which might be more intuitive

| Index | resourceGroupResource.GetServiceBusNamespace(string servicebusNamespaceName); |
| Add/Update | resourceGroupResource.GetServiceBusNamespaces().CreateOrUpdate(Azure.WaitUntil waitUntil, string name, ServiceBusNamespaceData data) |

> | Contains | resourceGroupResource.GetServiceBusNamespaces(). Exists(string servicebusNamespaceName) |

Reviewer 2: ```suggestion
| Contains | resourceGroupResource.GetServiceBusNamespaces().Exists(string servicebusNamespaceName) |
```

> | Delete |  resourceGroupResource.GetServiceBusNamespace(string servicebusNamespaceName).Delete() or armClient.GetServiceBusQueueResource(ResourceIdentifior resourceIdentifior).Delete()|

Reviewer 2: ```suggestion
| Delete |  resourceGroupResource.GetServiceBusNamespace(string servicebusNamespaceName).Delete() or armClient.GetServiceBusQueueResource(ResourceIdentifior resourceIdentifier).Delete()|
```
and could we put the second invocation at the front? It has fewer REST API calls.
We are planning to support a method on the collection to "delete a resource by name", therefore you will be able to do this like `resourceGroupResource.GetServiceBusNamespaces().Delete(string name)`. But now we do not have that yet

Author: Is this new method under discussion with archboard?

Author: > and could we put the second invocation at the front? It has fewer REST API calls. We are planning to support a method on the collection to "delete a resource by name", therefore you will be able to do this like `resourceGroupResource.GetServiceBusNamespaces().Delete(string name)`. But now we do not have that yet

Is this new method under discussion with archboard?

> Remember, all the Azure resource, including the resource group itself, can be managed by using the similar code above. All you need to do is find out the correct Azure management sdk packages and important them in your project. Package names usually look like this: `Azure.ResourceManager.[ResourceProviderName]`.

Reviewer 2: ```suggestion
Remember, all the Azure resource, including the resource group itself, can be managed by using the similar code above. All you need to do is to find out the correct Azure management SDK packages and import them in your project. Package names usually look like this: `Azure.ResourceManager.[ResourceProviderName]`.
```

> To learn more about the ResourceIdentifier, refer to [Structured Resource Identifier](#structured-resource-identifier)

Reviewer 3: ```suggestion
To learn more about the ResourceIdentifier, please refer to [Structured Resource Identifier](#structured-resource-identifier)
```

For more information about the `Azure.Identity.DefaultAzureCredential` class, see [DefaultAzureCredential Class](/dotnet/api/azure.identity.defaultazurecredential).

### Management SDK Cheat Sheet

To get started with the Azure management SDK for .NET, imagine you have a task to create/list/update/delete a typical Azure service bus namespace, follow these steps:

1. Authenticate to the subscription and resource group that you want to work on.

```csharp
using Azure.Identity;
using Azure.ResourceManager;
using Azure.ResourceManager.ServiceBus;

var armClient = new ArmClient(new DefaultAzureCredential());
SubscriptionResource subscription = armClient.GetDefaultSubscription();
ResourceGroupResource resourceGroupResource = 
    client.GetDefaultSubscription().GetResourceGroup(resourceGroupName);
```

> 2. Find the corresponding method to manage your Azure resource

Reviewer: ```suggestion
1. Find the corresponding method to manage your Azure resource.
```

To accomplish this, we're introducing three standard types for all resources in Azure:

> #### **[ResourceName]Resource.cs**

Reviewer 2: Should we use "class" to introduce this? Something.cs is like to introduce the source code instead of the library itself
```suggestion
#### **[ResourceName]Resource** class
```

Author: agree

> #### **[ResourceName]Resource.cs**

Author: We'd like to keep this section naming because customer can understand the naming convention of azure sdk

> #### **[ResourceName]Resource.cs**

Author: > 

We'd like to keep this section naming because customer can understand the naming convention of azure sdk

This type represents a full resource client object that contains a **Data** property exposing the details as a **[ResourceName]Data** type.
It also has access to all of the operations on that resource without needing to pass in scope parameters such as subscription ID or resource name. This makes it convenient to directly execute operations on the result of list calls, since everything is returned as a full resource client now.

```csharp
ArmClient armClient = new ArmClient(new DefaultAzureCredential());
string rgName = "myResourceGroup";
SubscriptionResource subscription = await armClient.GetDefaultSubscriptionAsync();
ResourceGroupCollection rg = await subscription.GetResourceGroups().GetAsync(rgName);

> await foreach (VirtualMachineResource vm in rg.GetVirtualMachines().GetAllAsync())

Reviewer 2: The other part of code is mainly using the synchronous method, should we align?

{
    //previously we would have to take the resourceGroupName and the vmName from the vm object
    //and pass those into the powerOff method as well as we would need to execute that on a separate compute client

>     await vm.StartPowerOff().WaitForCompletionAsync();

Reviewer 2: This is old code. We should use 
```suggestion
    await vm.PowerOffAsync(WaitUntil.Completed);
```

}
```

> #### **[ResourceName]Data.cs**

Reviewer: ```suggestion
#### Resource data types
```

|-|-|
| Iterate/List | GetAll() |
| Index | Get(string name) |
| Add | CreateOrUpdate(Azure.WaitUntil waitUntil, string name, [ResourceName]Data data) |
| Contains | Exists(string name) |

> For most things, the parent will be a **ResourceGroup**. However, each parent-child relationship is represented this way. For example, a **Subnet** is a child of a **VirtualNetwork** and a **ResourceGroup** is a child of a **Subscription**.

Reviewer 3: The word "however" here is confusing. Because the first half says "most" cases, combined with the last sentence, I think here should be only still exist special cases? But I don't see a clear transitive relationship.

var armClient = new ArmClient(new DefaultAzureCredential());

// Next we get a resource group object
// ResourceGroup is a [ResourceName] object from above
SubscriptionResource subscription = await armClient.GetDefaultSubscriptionAsync();

> ResourceGroupResource resourceGroup = (await subscription.GetResourceGroups().GetAsync("myRgName").Value);

Reviewer 2: The Value might be redundant and is the parenthesis matching with each other? It looks not correct

> SubscriptionResource subscription = await armClient.GetDefaultSubscriptionAsync();
> ResourceGroupResource resourceGroup = (await subscription.GetResourceGroups().GetAsync("myRgName").Value);

Reviewer: ```suggestion
SubscriptionResource subscription =
    await armClient.GetDefaultSubscriptionAsync();
ResourceGroupResource resourceGroup =
    await subscription.GetResourceGroups().GetAsync("myRgName");
```

// Next we get the collection for the virtual machines
// vmCollection is a [ResourceName]Collection object from above
VirtualMachineCollection vmCollection = resourceGroup.GetVirtualMachines();

// Next we loop over all vms in the collection
// Each vm is a [ResourceName] object from above
await foreach(VirtualMachineResource vm in vmCollection.GetAllAsync())
{
    // We access the [ResourceName]Data properties from vm.Data
    if(!vm.Data.Tags.ContainsKey("owner"))
    {

>         // We can also access all operations from vm since it is already scoped for us

Reviewer 2: The below line is using the old API, which should be 
await vm.AddTagAsync("owner", GetOwner());

}
 ```

### Structured Resource Identifier

> Resource IDs contain useful information about the resource itself, but they're plain strings that have to be parsed. Instead of implementing your own parsing logic, you can use a `ResourceIdentifier` object that will do the parsing for you: `new ResourceIdentifier("myid");`.

Reviewer 3: ```suggestion
Resource IDs contain useful information about the resource itself, but they're plain strings that have to be parsed. Instead of implementing your own parsing logic, you can use a `ResourceIdentifier` object that will do the parsing for you.
```
A detailed sample is provided below, I don't think this is needed here anymore.

However, keep in mind that some of those properties could be null. You can usually tell by the ID string itself which type a resource ID is. But if you're unsure, check if the properties are null, or use the `Try` methods to retrieve the values, as shown in the following example.

#### Example: Resource Identifier Generator

```csharp

> ResourceIdentifior resourceId = AvailabilitySetResource.CreateResourceIdentifier("aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee","resourceGroupName","resourceName"); 

Reviewer 2: ```suggestion
ResourceIdentifior resourceId = AvailabilitySetResource.CreateResourceIdentifier("aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee", "resourceGroupName", "resourceName"); 
```

```

> You may not want to manually create the resourceId from a pure string, each `[ResourceName]Resource` class has a static method that can help you create the resource identifier string

Reviewer 2: Should this line appear before the code example above?

However, keep in mind that some of those properties could be null. You can usually tell by the ID string itself which type a resource ID is. But if you're unsure, check if the properties are null, or use the `Try` methods to retrieve the values, as shown in the following example.

#### Example: Resource Identifier Generator

You may not want to manually create the `resourceId` from a pure `string`. Each `{ResourceName}Resource` class has a static method that can help you create the resource identifier string.

```csharp
ResourceIdentifier resourceId = 
    AvailabilitySetResource.CreateResourceIdentifier(
        "aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee",
        "resourceGroupName", 
        "resourceName"); 
```

> #### Example: ResourceIdentifier TryGet methods

Reviewer: ```suggestion

#### Example: ResourceIdentifier TryGet methods
```

Console.WriteLine($"Subnet: {id.Name}");
```

> #### Managing Existing Resources By ID

Reviewer: ```suggestion
#### Manage existing resources
```

> Performing operations on resources that already exist is a common use case when using the management client libraries. In this scenario, you usually have the identifier of the resource you want to work on as a string. Although the new object hierarchy is great for provisioning and working within the scope of a given parent, it is not the most efficient when it comes to this specific scenario.  

Reviewer 3: ```suggestion
Performing operations on resources that already exist is a common use case when using the management client libraries. In this scenario, you usually have the ID of the resource you want to work on as a string. Although the new object hierarchy is great for provisioning and working within the scope of a given parent, it is not the most efficient when it comes to this specific scenario.  
```

> Here's an example how you to access an `AvailabilitySetResource` object and manage it directly with its ID:

Reviewer 3: ```suggestion
Here's an example how you to access an `AvailabilitySetResource` object and manage it directly with its resource ID:
```

// Code omitted for brevity

> ResourceIdentifior resourceId = AvailabilitySetResource.CreateResourceIdentifier("aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee","resourceGroupName","resourceName"); 

Reviewer 2: ```suggestion
ResourceIdentifior resourceId = AvailabilitySetResource.CreateResourceIdentifier("aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee", "resourceGroupName", "resourceName"); 
```

// We construct a new armClient to work with
ArmClient armClient = new ArmClient(new DefaultAzureCredential());
// Next we get the specific subscription this resource belongs to
SubscriptionResource subscription = await armClient.GetSubscriptions().GetAsync(resourceId.SubscriptionId);
// Next we get the specific resource group this resource belongs to

> ResourceGroupResource resourceGroup = (await subscription.GetResourceGroups().GetAsync(resourceId.ResourceGroupName)).Value;

Reviewer 2: like the subscription line above, we do not need the `.Value` here

// Finally we get the resource itself
// Note: for this last step in this example, Azure.ResourceManager.Compute is needed

> AvailabilitySetResource availabilitySet = (await resourceGroup.GetAvailabilitySets().GetAsync(resourceId.Name)).Value;

Reviewer 2: Same

```

> This approach required a lot of code and three API calls to Azure. The same can be done with less code and without any API calls by using extension methods that we've provided on the client itself. These extension methods allow you to pass in a resource identifier and retrieve a scoped resource client. The object returned is a [ResourceName]Resource(#resourcenameresourcecs). Since it hasn't reached out to Azure to retrieve the data yet, the `Data` property will be null.

Reviewer 2: Is this `[ResourceName]Resource(#resourcenameresourcecs)` supposed to be a link? It is not in correct format here

> This approach required a lot of code and three API calls to Azure. The same can be done with less code and without any API calls by using extension methods that we've provided on the client itself. These extension methods allow you to pass in a resource identifier and retrieve a scoped resource client. The object returned is a [ResourceName]Resource(#resourcenameresourcecs). Since it hasn't reached out to Azure to retrieve the data yet, the `Data` property will be null.

Reviewer 2: And why is the resource client "scoped"?
```suggestion
This approach required a lot of code and three API calls to Azure. The same can be done with less code and without any API calls by using extension methods that we've provided on the client itself. These extension methods allow you to pass in a resource identifier and retrieve a resource client. The returned object is a [ResourceName]Resource(#resourcenameresourcecs). Since it hasn't reached out to Azure to retrieve the data yet, the `Data` property will be null.
```
The last sentence is not very accurate - the field `_data` is null indeed but if you call the property `Data`, you will get an exception.

Reviewer 2: We may need to note our customers, in this case, you need to call `Get` or `GetAsync` on the resource so that you could use the Data property.
And we have a `HasData` property to tell you if this resource instance contains a data

Author: anyway we can return null instead of throw exception when calling a Data when the get method is not called?

So, the previous example would end up looking like this:

```csharp

> ResourceIdentifior resourceId = AvailabilitySetResource.CreateResourceIdentifier("aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee","resourceGroupName","resourceName"); 

Reviewer 2: ```suggestion
ResourceIdentifior resourceId = AvailabilitySetResource.CreateResourceIdentifier("aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee", "resourceGroupName", "resourceName"); 
```

// We construct a new armClient to work with

> ArmClient armClient = new ArmClient(new DefaultAzureCredential());

Reviewer: ```suggestion
var armClient = new ArmClient(new DefaultAzureCredential());
```

// Next we get the AvailabilitySet resource client from the armClient
// The method takes in a ResourceIdentifier but we can use the implicit cast from string
AvailabilitySetResource availabilitySet = armClient.GetAvailabilitySet(resourceId);
// At this point availabilitySet.Data will be null and trying to access it will throw
// If we want to retrieve the objects data we can simply call get

> availabilitySet = (await availabilitySet.GetAsync()).Value;

Reviewer 2: The `.Value` is redundant here

> availabilitySet = (await availabilitySet.GetAsync()).Value;

Reviewer: ```suggestion
availabilitySet = await availabilitySet.GetAsync();
```

// we now have the data representing the availabilitySet
Console.WriteLine(availabilitySet.Data.Name);
```

### Check if a Resource exists

If you aren't sure if a resource you want to get exists, or you just want to check if it exists, you can use `Exists()` or `ExistsAsync()` methods, which can be invoked from any [ResourceName]Collection class.

> `Exists()` returns a `Response<bool>` while `ExistsAsync()` as its async version returns a `Task<Response<bool>>`, in the `Reponse<bool>` object, you can visit its `Value` property to check if a [ResourceName] exists. The `Value` is `false` if the resource does not exist and vice versa.

Reviewer 3: ```suggestion
`Exists()` returns a `Response<bool>` while `ExistsAsync()` as its async version returns a `Task<Response<bool>>`. In the `Reponse<bool>` object, you can visit its `Value` property to check if a [ResourceName] exists. The `Value` is `false` if the resource does not exist and vice versa.
```

In previous versions of packages, you would have to catch the `RequestFailedException` and inspect the status code for 404. With this new API, we hope that this can boost the developer productivity and optimize resource access.

```csharp
ArmClient armClient = new ArmClient(new DefaultAzureCredential());
SubscriptionResource subscription = await armClient.GetDefaultSubscriptionAsync();
string rgName = "myRgName";

try
{

>     ResourceGroupResource myRG = (await subscription.GetResourceGroups().GetAsync(rgName)).Value;

Reviewer 2: `.Value` is redundant

    // We can get the resource group now that we know it exists.
    // This does introduce a small race condition where resource group could have been deleted between the check and the get.

>     ResourceGroupResource myRG = (await subscription.GetResourceGroups().GetAsync(rgName)).Value;

Reviewer 2: same

// First, initialize the ArmClient and get the default subscription
ArmClient armClient = new ArmClient(new DefaultAzureCredential());
// Now we get a ResourceGroup collection for that subscription
SubscriptionResource subscription = await armClient.GetDefaultSubscriptionAsync();
ResourceGroupCollection rgCollection = subscription.GetResourceGroups();

// With the collection, we can create a new resource group with an specific name
string rgName = "myRgName";
Location location = Location.WestUS2;
ResourceGroupData rgData = new ResourceGroupData(location);

> ResourceGroupResource resourceGroup = (await rgCollection.CreateOrUpdateAsync(rgName, rgData)).Value;

Reviewer 2: This `.Value` is mandatory, because `ArmOperation<T>` does not have an implicit operator

Author: Are we planning to add implicit convert for ArmOperation like what we did for Azure.Response<T>? this inconsistency will confuse our customer

Author: > This `.Value` is mandatory, because `ArmOperation<T>` does not have an implicit operator

Are we planning to add implicit convert for ArmOperation like what we did for Azure.Response? this inconsistency will confuse our customer

```

### List all resource groups

```csharp
// First, initialize the ArmClient and get the default subscription
ArmClient armClient = new ArmClient(new DefaultAzureCredential());
SubscriptionResource subscription = await armClient.GetDefaultSubscriptionAsync();
// Now we get a ResourceGroup collection for that subscription
ResourceGroupCollection rgCollection = subscription.GetResourceGroups();
// With GetAllAsync(), we can get a list of the resources in the collection

> await foreach (ResourceGroupResource rg in rgCollection.GetAllAsync())

Reviewer 2: We could omit the `GetAllAsync()` call here because the collection itself implements `IAsyncEnumerable<T>`

```csharp
// Note: Resource group named 'myRgName' should exist for this example to work.
ArmClient armClient = new ArmClient(new DefaultAzureCredential());
SubscriptionResource subscription = await armClient.GetDefaultSubscriptionAsync();
string rgName = "myRgName";

> ResourceGroupResource resourceGroup = (await subscription.GetResourceGroups().GetAsync(rgName)).Value;

Reviewer 2: `.Value` could be omitted

resourceGroup = await resourceGroup.AddTagAsync("key", "value");
```

### Delete a resource group

```csharp
ArmClient armClient = new ArmClient(new DefaultAzureCredential());
SubscriptionResource subscription = await armClient.GetDefaultSubscriptionAsync();
string rgName = "myRgName";

> ResourceGroupResource resourceGroup = (await subscription.GetResourceGroups().GetAsync(rgName)).Value;

Reviewer 2: Same
